app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt

```diff
@@ -0,0 +1,140 @@
+                    }
+                    MovieCategory.NOW_PLAYING_TV -> {
+                        movieService.fetchTopRatedMovies()
+                    }
+                    MovieCategory.UPCOMING_TODAY -> {
+                        movieService.fetchUpcomingMovies()
+                    }
+                    MovieCategory.UPCOMING_THIS_WEEK -> {
+                        movieService.fetchTopRatedMovies()
+                    }
+                }
+                emit(movieResponse.movies)
+            }.flatMapLatest { apiMovies ->
+                movieDao.favorites()
+                    .map { favoriteMovies ->
+                        apiMovies.map { apiMovie ->
+                            apiMovie.toMovie(isFavorite = favoriteMovies.any { it.id == apiMovie.id })
+                        }
+                    }
+            }.shareIn(
+                scope = CoroutineScope(bgDispatcher + CoroutineExceptionHandler { _, throwable ->
+                    throwable.printStackTrace()
+                }),
+                started = SharingStarted.WhileSubscribed(1000L),
+                replay = 1,
+            )
+        }
+
+    private val favorites = movieDao.favorites().map {
+        it.map { dbFavoriteMovie ->
+            Movie(
+                id = dbFavoriteMovie.id,
+                imageUrl = dbFavoriteMovie.posterUrl,
+                title = "",
+                overview = "",
+                isFavorite = true,
+            )
+        }
+    }.shareIn(
+        scope = CoroutineScope(bgDispatcher),
+        started = SharingStarted.WhileSubscribed(1000L),
+        replay = 1,
+    )
+
+    override fun movies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    override fun movieDetails(movieId: Int): Flow<MovieDetails> = flow {
+        emit(movieService.fetchMovieDetails(movieId) to movieService.fetchMovieCredits(movieId))
+    }.flatMapLatest { (apiMovieDetails, apiMovieCredits) ->
+        movieDao.favorites()
+            .map { favoriteMovies ->
+                apiMovieDetails.toMovieDetails(
+                    isFavorite = favoriteMovies.any { it.id == apiMovieDetails.id },
+                    crew = apiMovieCredits.crew,
+                    cast = apiMovieCredits.cast
+                )
+            }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    override suspend fun addMovieToFavorites(movieId: Int) {
+        val movie = findMovie(movieId)
+        movieDao.insertMovie(
+            DbFavoriteMovie(
+                id = movie.id,
+                posterUrl = movie.imageUrl
+            )
+        )
+    }
+
+    private suspend fun findMovie(movieId: Int): Movie {
+        return moviesByCategory.values.asFlow().flattenMerge()
+            .filter { it -> it.firstOrNull { it.id == movieId } != null }
+            .first().first { it.id == movieId }
+    }
```

Comment:
You can just use `moviesByCategory.values.forEach` instead of this `.asFlow().flattenMerge()` and then `.first()`

Reply:
Don't know what exactly you mean, how am I supposed to retrieve Movie using forEach?

Reply:
It would look something like this:
```
moviesByCategory.values.forEach {
            val movie = it.first().firstOrNull { movie -> movie.id == movieId }
            if (movie != null) return movie
}
```